docs/src/_docs/guide.md

```diff
@@ -143,15 +144,84 @@ Assign the function without inputs to a variable, and Turing will produce a samp
 
 ```julia
 g = gdemo()
-g
-# Output: (0.685690547873451, -1.1972706455914328)
+```
+Output:
+```
+(0.685690547873451, -1.1972706455914328)
+```
+
+#### Generating Vectors of Quantities
+
+More broadly, Turing models can be treated as generative by providing default values in the model declaration, and then calling that model without parameters.
```

Comment:
default values can only be `missing` at the moment, any other value will be treated as missing values. 

@mohamed82008 is the above summary accurate?

Reply:
It is kind of accurate yes. Default values currently can be anything but only the size and type information of the default value will be used if it is a vector. The value itself is not used anywhere. When the input is a vector (or matrix, etc), it is important to have a default input for the case when it is treated as a parameter to know the eltype and size to pre-allocate it. When the input is a scalar, the default value is useless, as we don't need to pre-allocate. I would just add more explanation explaining why this feature exists. Also perhaps change "calling that model without parameters" -> "calling that model without arguments" to avoid confusion with parameters and data variables.

Reply:
Are these all valid defaults? 

```julia
zeros(10)
Vector{Real}(undef, 10)
Vector{Union{Missing, Real}}([missing, missing])
Vector{Union{Missing, Real}}([missing, 2])
[1,2,3,9,10,16]
```

My proposed addition to explain this behavior:

> #### What to Use as a Default Value
> 
> Currently, the actual _value_ of the default argument does not matter. Only the dimensions and type of a non-atomic value are relevant. Turing uses default values to pre-allocate vectors when they are treated as parameters, because if the value is not provided, the model will not know the size or type of a vector. Consider the following model:
> 
> ```julia
> @model generator(x) = begin
>   s ~ InverseGamma(2,3)
>   m ~ Normal(0,sqrt(s))
>   for i in 1:length(x)
>       x[i] ~ Normal(m, sqrt(s))
>   end
>   return s, m
> end
> ```
> 
> If we are trying to generate random random values from the `generator` model and we call `sample(generator(), HMC(1000, 0.01, 5))`, we will receive an error. This is because there is no way to determine `length(x)`, whether `x` is a vector, and the type of the values in `x`.
> 
> A sensible default value might be:
> 
> ```julia
> @model generator(x = zeros(10)) = begin
>   s ~ InverseGamma(2,3)
>   m ~ Normal(0,sqrt(s))
>   for i in 1:length(x)
>       x[i] ~ Normal(m, sqrt(s))
>   end
>   return s, m
> end
> ```
> 
> In this case, the model compiler can now determine that `x` is a `Vector{Float64,1}` of length 10, and the model will work as intended. It doesn't matter what the values in the vector are — at current, `x` will be treated as a parameter if it assumes its default value, i.e. no value was provided in the function call for that variable.

Reply:
Yes these are valid defaults with one restriction. Variables used in Hamiltonian samplers need to have `Real` `eltype`. The last case is integer, so it can only be used for integer variables, which are obviously not Hamiltonian-compatible.

Reply:
When sampling from prior, we don't need the `Real` restriction.

Reply:
> When sampling from prior, we don't need the `Real` restriction.

What are the restrictions on there if `Real` is not required?

Reply:
Its `eltype` needs to admit the `eltype` of the variable, e.g. `Int` or `Float64`.

Reply:
So would it be fair to say the following:

> The type of the vector should be either `Int` or `Float64`, depending on which sampler is being used. A Hamiltonian sampler like `HMC` or `NUTS` requires values to be of type `Float64`, whereas non-Hamiltonian samplers like `SMC` or `PG` permit the use of `Int` variables.

Reply:
I think the following is better: 

"The element type of the vector (or matrix) should match the type of the random variable, `<: Integer` for discrete random variables and `<: AbstractFloat` for continuous random variables. Moreover, if the continuous random variable is to be sampled using a Hamiltonian sampler, the vector's element type needs to be `Real` to enable auto-differentiation through the model which uses special number types that are sub-types of `Real`. Finally, when using a particle sampler, a `TArray` should be used."

These are general tips when pre-allocating a vector of parameters, regardless of whether or not, it is a "default value".

Reply:
Good one, threw that into the guide.